importSVG Overhaul

aprospero opened the pull request:
### Abstract
The builtin svg import functionality, which is part of Draft workbench, could be more error tolerant, could support cuts in shapes and should produce paths that are easy to process.

Therefore this PR wants to add:

1. robustness against non-matching start- and endpoints due to accumulated rounding errors when using relative coordinates
2. better filtering of invalid or nonsensical data, eg. zero-length or microscopic short edges 
3. optional cuts of sub-paths enclosing other sub-paths 
4. apply matrix transformations while keeping degeneration of created edges as low as possible.

To achieve that the import process is segmented in several phases:

1. init
2. parse
3. optimize/correct
4. create faces
5. cut (optional)

Phase 2, 'parse', creates a set of metadata for each path segment. Phase 3 optimizes and corrects the metadata, while Phase 4 creates Faces. from the corrected metadata.

Following a short introduction to the four PR subjects.

### Robustness for matching start- and endpoints
The huge majority of paths people want to import are closed paths. Unfortunately there are many ways to create paths in svg that look like they are closed, while mathematically spoken they are not. And even the most careful svg creator may end up with a file that defines both, start and endpoint on almost the same spot, but differing on a minimal scale.
This can be alleviated by using a precision attribute controlling how many fractional digits are considered for identity and geometric tests.

### Filtering of invalid or nonsensical data
This PR divides the svg parsing process in several phases. First a list of metadata about sub-paths is generated from svg. Already in this early process too tiny edges are filtered. 
But there are special cases, like cubic beziers that can have zero distance between start and end while maintaining full validity. For this reason detecting invalid or nonsensical data is also part of the following edge generating phase.  Eg. if beziers have no curve they are replaced with line segments. 
The last optimization for each closed wire is correcting start- and endpoint if they are supposed to be identical but aren't.

### Optional cuts of sub-paths enclosing other sub-paths 
Often svg paths have holes or cutouts. They follow the svg fill parameter and its [fill-rule property](https://www.w3.org/TR/SVG11/painting.html#FillRuleProperty). This PR adds the option to cut paths following the evenodd svg fill rule (see link).
Unfortunately evenodd was the low hanging fruit but not the default for svg objects. So there will be complaints about differences between svg output and freeCAD import. 
Implementing the default fill rule (nonzero) will (hopefully) be added in a later PR.

### Apply matrix transformations while avoiding degeneration
Paths generated with svg import often break standard procedures like champfer or fillet. That's due to the fact that any imported path is subject to a matrix transformation that is applied in a way that produces edge degeneration. Eg. straight lines degenerate to straight first order bsplines.
In 2015 there was a [Bug Report](https://tracker.freecad.org/view.php?id=2062) complaining about differences in the size of generated objects compared to its size defined in the svg file. Long story short, the bug report suggests that the svg file was ambiguous/broken, which resulted in a wrong transformation applied. Despite that on behalf of this bug report a cruical piece of code got reverted which introduced unnecessary degenerations to all generated paths. Eg. [this Issue](https://github.com/FreeCAD/FreeCAD/issues/20264) is most possibly result of the mentioned degeneration.
This PR reactivates and 'refreshes' the code that avoids degenerations, but was disabled 2015.

aprospero commented:
@phaseloop fyi

hyarion commented:
Regarding precision, OCCT have support for fixing wires using `ShapeFix_Wire`, you might want to check it out as it can fix a lot of other issues as well (small segments, intersections etc).

mnesarco commented:
@aprospero your changes in SvgWorkbench were excellent. Thank you. Are these changes the same?

hyarion commented:
I can neither see it used in the new, nor old implementation but I do see that you are trying to do something similar when you are rounding the coordinates. When you do that you always introduce an error.

https://dev.opencascade.org/doc/refman/html/class_shape_fix___wire.html there is the c++ api and python api is built from it using swig. So I hope it is available.

In this you can set precision and tolerance that you want to use (both like setTolerence() on the object and in each fix-call if you need a specific)

hyarion commented:
Sorry, I forgot that python can't access OCCT directly.
The functionality is exposed through `Part.ShapeFix.Wire()` though.

There's a test which shows how it can be used here: https://github.com/FreeCAD/FreeCAD/blob/main/src/Mod/Part/TestPartApp.py#L760-L896

And you have the bindings available here: https://github.com/FreeCAD/FreeCAD/blob/main/src/Mod/Part/App/ShapeFix/ShapeFix_Wire.pyi

I don't know if there's any new functionality in occt that isn't yet in the python bindings, but let me know if there is and you want me to help you adding it.

Btw, thanks for looking into the svg import, it's much appreciated!

aprospero commented:
> There's a test which shows how it can be used here: https://github.com/FreeCAD/FreeCAD/blob/main/src/Mod/Part/TestPartApp.py#L760-L896
> 
> And you have the bindings available here: https://github.com/FreeCAD/FreeCAD/blob/main/src/Mod/Part/App/ShapeFix/ShapeFix_Wire.pyi
> 
Thanks for pointing me in the right direction - you're correct, this mechanism hasn't been used yet for svg import. I have confused it with face.fix().

I'm having a look at it right now but I have a hard time figuring out how to initialize that beast. If I understand it right I have to provide the wire and a face, right? So how am I supposed to provide a face out of thin air? If I would have to provide a surface I would choose the xy-plane, but a face?

Does anyone  have a hint?

aprospero commented:
> @aprospero your changes in SvgWorkbench were excellent. Thank you. Are these changes the same?

@mnesarco They are in part based on the insights I aquired working on your svgWorkbench import code. 

With this update a new Python module will be added to the Draft WB which you could use to replace the module in your workbench that I created for the same purpose. This way you would always have the latest fixes from the main repo.
Additionally, there is not much code in the  original importSVG.py file that has changed. Most important to me seems the change in applying Transformations. It should not be much work to transfer that into your svgWorkbench. 
But I thought you are considering using a library? Maybe the project would benefit when there are two different kinds of svg import tools available?

mnesarco commented:
> With this update a new Python module will be added to the Draft WB which you could use to replace the module in your workbench that I created for the same purpose. This way you would always have the latest fixes from the main repo. Additionally, there is not much code in the original importSVG.py file that has changed. Most important to me seems the change in applying Transformations. It should not be much work to transfer that into your svgWorkbench. But I thought you are considering using a library? Maybe the project would benefit when there are two different kinds of svg import tools available?

I plan to keep SvgWorkbench code separated from upstream FreeCAD so users can get updates without requiring them to install newer versions of FreeCAD, it is not uncommon that newer versions has more bugs than former ones.

It would be great to have a set of svg files for testing so I can check if things diverge too much over time...

aprospero commented:
> It would be great to have a set of svg files for testing so I can check if things diverge too much over time...

Yeah, that would really be great. Someone who could do some testing and generating test files. 
It's always a good idea to let other do the testing - the engineer who built it will always test the use cases he already took care of. I already have guys at hand who do tests - but on a completely voluntarily level. Maybe you could provide some test files?

phaseloop commented:
I'm AFK for a week but I'm happy to run it on my set of svg torture files.

If anyone wants to test it too - freepik.com have millions of free svg files to use.

hyarion commented:
> Thanks for pointing me in the right direction - you're correct, this mechanism hasn't been used for svg import yet. I have confused it with face.fix().
> 
> I'm having a look at it right now but I have a hard time figuring out how to initialize that beast. If I understand it right I have to provide the wire and a face OR a surface, right? So how am I supposed to provide a face out of thin air? If I would have to provide a surface I would choose the xy-plane. But in that case there is no matching init function and loading the Analyzer doesn't seem to be properly explained in the docs.

Took some time but I think I've figured out the API buy looking at the `ShapeFix_Wire.pyi` and `ShapeFix_WirePyImp.cpp` files.

You can call it without arguments. To my understanding, the face might just be a surface which your points will be placed on if moved.

The following example is based on a wire from a sketch:
<img width="238" alt="image" src="https://github.com/user-attachments/assets/9851967c-6794-4c03-8fd1-395c96afb770" />

This seems like it works based on the python console output:
```py
>>> wire = shp  # <-- I sent the shape to the python console before this call
>>> wire
<Wire object at 0x60000f8f73e0>
>>> wire.isClosed()  # <-- check that it *isn't* closed at this point
False
>>> import Part
>>> fix = Part.ShapeFix.Wire()
>>> fix.load(wire)
>>> fix.Precision = 1.5  # Note: in mm, not number of digits
>>> fix.fixClosed()
True
>>> w = fix.wire()
>>> w
<Wire object at 0x60000f8f0500>
>>> w.isClosed()  # <-- check that it now *is* closed at this point
True
```

So in your case it would be something like:
```py
import Part
fix = Part.ShapeFix.Wire()
fix.load(wire)
fix.Precision = 10**precisionDigits
fix.fixClosed()
fixedWire = fix.wire()
```

There are ofc other fixing operations available too, depending on your needs.

aprospero commented:
> Took some time but I think I've figured out the API buy looking at the `ShapeFix_Wire.pyi` and `ShapeFix_WirePyImp.cpp` files.

Cool, thank you! That's exactly the glue I needed. 

> 
> To my understanding, the face might just be a surface which your points will be placed on if moved.

Yeah, and that's the xy-plane in all svg import cases by default. Beware of confusing surface with face! A plane is a surface but not a face.

> 
> So in your case it would be something like:
> 
> ```python
> import Part
> fix = Part.ShapeFix.Wire()
> fix.load(wire)
> fix.Precision = 10**precisionDigits
> fix.fixClosed()
> fixedWire = fix.wire()
> ```

Thanks a lot, that's exactly the piece I needed.

> 
> There are ofc other fixing operations available too, depending on your needs.

I know.... ;)

luzpaz commented on src/Mod/Draft/SVGPath.py:
```diff
@@ -0,0 +1,788 @@
+        
+        # for cbeziers we also relocate the second pole
+        if pds["type"] == "cbezier":
+                pds["pole2"] = pds["pole2"].add(delta)
```
bad indentation

aprospero replied:
Thanks for having a look. My last commit history streamlining seems to have introduced some wear....

luzpaz commented on src/Mod/Draft/SVGPath.py:
```diff
@@ -0,0 +1,788 @@
+                pds["pole2"] = pds["pole2"].add(delta)
+        # for qbeziers we also relocate the pole by half of the delta
+        elif pds["type"] == "qbezier":
+                pds["pole"] = pds["pole"].add(delta.scale(0.5, 0.5, 0))
```
bad indentation

aprospero commented:
> So in your case it would be something like:
> 
> ```python
> import Part
> fix = Part.ShapeFix.Wire()
> fix.load(wire)
> fix.Precision = 10**precisionDigits
> fix.fixClosed()
> fixedWire = fix.wire()
> ```
> 
> There are ofc other fixing operations available too, depending on your needs.

@hyarion  thanks again for putting me on track to this feature - I have played with it for a while now and it could improve import results in some corner cases. 

On the other hand it is so badly documented I simply don't know if I use it right. Eg. one has to provide a surface or a face - without it the fixers state never gets 'ready' - but even with provided xy-plane as surface many problematic edges simply fall through and later on fail while creating a face.

Also I think the whole error processing as I implemented it doesn't take much time, is transparent and therefore adapted to the use-case, makes educated guesses based on the knowledge of the raw svg data and does NOT introduce unintended errors. 

Therefore I would prefer to add the ShapeFix.Wire as a post-process instead to replace the existing error correction. The better the Wires coming from the import the better the results the fixer can deliver. Since I have not much experience with the fixer it's better to take the time and get a better grip on it than to hurry and to push it into this pull request by force. I.e. I would later come up with another PR adding the fixer as post-process.

What do you guys think?

Another thing, @hyarion, is your offer to show me how to add OCCT-Features to the python API to the python API. Maybe you can show me on the example for ShapeAnalysis_Wire? It's a c++ object that's also used by the fixer and can perform (you no say) analyzing tasks on the wire. There is an access function in the ShapeFix_wire for its underlying Analyzer, but unfortunately it's not exposed in Python and then would still miss the object definition for python, to access the Analyzers functionality.
That would allow me to see the test results on which the fixer bases its actions...

Could you give me a hint - again?

hyarion commented:
> On the other hand it is so badly documented I simply don't know if I use it right. Eg. one has to provide a surface or a face - without it the fixers state never gets 'ready' - but even with provided xy-plane surface many problematic edges simply fall through and later on fail while creating a face.

Yeah, it was done a long time ago.

> Also I think the whole error processing as I implemented it doesn't take much time, is transparent and therefore adapted to the use-case, makes educated guesses based on the knowledge of the raw svg data and does NOT introduce unintended errors (precision is only used for Coordinate comparisons, not for calculations - ok, with one necessary exception).

The problem I see with your approach is that it always rounds coordinates, even when not needed (if I understood your code correctly). While it is quick and easy, it does introduce issues.
One concrete issue is if you have a line like `y=kx` and you have segments through the points `x=[0,1,2,3,5,7]` then the resulting line would get a bit jagged depending on your rounding.

However I do acknowledge that your approach is easier to understand and work with due to the bad state that `ShapeFix_Wire` is in. So there's both pros and cons about both ways.

> Therefore I would prefer to add the ShapeFix.Wire as a post-process instead replacing the existing error correction. The better the Wires coming from the import the better the results the fixer can deliver. Since I have not much experience with the fixer it's better to take the time and get a better grip on it than to hurry and to push it into this pull request by force. I.e. I would later come up with another PR adding the fixer as post-process.
> 
> What do you guys think?

I agree, lets keep this as you propose and we can worry about OCCT fixer later.

> Another thing, @hyarion, is your offer to show me how to add OCCT-Features to the python API. Maybe you can show me on the example for ShapeAnalysis_Wire? It's a c++ object that's also used by the fixer and can perform (you no say) analyzing tasks on the wire. There is an access function in the ShapeFix_wire for its underlying Analyzer, but unfortunately it's not exposed in Python and then would still miss the object definition to access the Analyzers functionality. That would allow me to see the test results on which the fixer bases its actions...
> 
> Could you give me a hint - again?

ShapeAnalysis_Wire doesn't seem to be available, yet.

If I where to add support for it and improve documentation and types on `ShapeFix_Wire`, is this something you would like to use? (if not we can take it when we're looking over the binding generation, whenever that happens)

aprospero commented:
> The problem I see with your approach is that it always rounds coordinates, even when not needed (if I understood your code correctly). While it is quick and easy, it does introduce issues. One concrete issue is if you have a line like `y=kx` and you have segments through the points `x=[0,1,2,3,5,7]` then the resulting line would get a bit jagged depending on your rounding.

Hmm,  I tried to make sure that the result is only changed in places that need repairing, and that is in the nature of things. 

Rounding is generally not used in the actual calculations. It is only applied for comparison of coordinates.

The only exception is the calculation of the viewport transformation. Here it can lead to minimal distortions if a non-uniform scaling was selected for the viewport (importSVG.py:652). What sense a non-uniform scaling could have in the context of CAD is not clear to me, and the benefit gained with this approach, namely that a simple viewport transformation does not lead to the degeneration of all line segments to 1st order BSplines, is worth it in my opinion.

The actual paths are constructed in exactly the same way as without rounding - at least that was the goal. If you find code that doesn't respect that please let me know!

My biggest concern lies in the fact that the repair mechanism doesn't take the psychological implications of shape changes into account. For example, if a small edge is removed, it makes a huge difference whether it makes a neighbouring B-spline slightly longer and/or more/less curvy, or whether a neighbouring horizontal or vertical straight line suddenly becomes skewed. 
But this again is in the nature of things (until the day AI will do the job - and then we'll have a whole new set of problems).

> If I where to add support for it and improve documentation and types on ShapeFix_Wire, is this something you would like to use? (if not we can take it when we're looking over the binding generation, whenever that happens)

I'd like to use it for Analysis before blindly trying repairs without knowing what it will do. I'm not exactly sure what to expect from the Analysis itself, how fine grained information will be, but I hope to better understand the actions taken by the ShapeFixer_Wire.
So, I'd really like to use it for the PR adding ShapeFixer_wire. On the other hand: it's far away from being urgent.

hyarion commented:
@aprospero
When merging multi-commit PRs, we typically either **squash** the commits or **rebase** and merge them.

If we’re keeping the individual commits, we ask that each commit compiles and works properly, so that tools like git bisect remain effective.
It’s also helpful to prefix commits (e.g., Draft: ) to make it easier to see which module is affected.

I noticed you’ve put effort into the commit messages, and I’d prefer not to squash them if we don’t have to, it’s great context to preserve.

Would you prefer to squash the commits, or keep them as-is?
And if you'd like to keep them, can you confirm that each commit builds and works correctly on its own?

Let me know how you'd like to proceed.

aprospero commented:
> If we’re keeping the individual commits, we ask that each commit compiles and works properly, so that tools like git bisect remain effective. It’s also helpful to prefix commits (e.g., Draft: ) to make it easier to see which module is affected.

unfortunately I introduced some indentation errors when I rewrote the git history before setting up the PR (I had a much more detailed history beforehand which is not always appreciated by everyone).
These are fixed now but bisecting would not work. 

Additional, even though I love git because of its ability to keep detailed information about each commits content and intent, there are many tools - even in the native git toolcase which ignore or even delete the comment sections of commit messages in the process, so this info possibly won't outlive eg. the next migration anyway. 

> I noticed you’ve put effort into the commit messages, and I’d prefer not to squash them if we don’t have to, it’s great context to preserve.

Most valuable info is certainly in the comments to [d6c9159](https://github.com/FreeCAD/FreeCAD/pull/20749/commits/d6c9159fcf978bd17c2b4c10833cd3a42d160e26) - I will cite it here and then I'd vote for a **squash merge**. I'll keep the history in my fork - if it's ever of any use.

Thanks again for caring, @hyarion!
